File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int l=nums.length;
        List<List<Integer>> ans = new ArrayList<>();
        for(int i=0; i<l; i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int lo=i+1, hi=l-1;
            while(lo<hi){
                int sum=nums[i]+nums[lo]+nums[hi];
                if(sum<0) lo++;
                else if(sum>0) hi--;
                else{
                    ans.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
                    int p=lo;
                    while(lo<l && nums[p]==nums[lo]) ++lo;
                    p=hi;
                    while(hi>=0 && nums[p]==nums[hi]) --hi;
                }
            }
        }
        return ans;
    }
}

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        i=0
        ans=[]
        l=len(nums)
        while i<l:
            lo, hi = i+1, l-1
            while lo<hi:
                if nums[lo]+nums[hi]==-1*nums[i]:
                    ans.append([nums[i], nums[lo], nums[hi]])
                    p=lo
                    while lo<l and nums[lo]==nums[p]: lo+=1
                    p=hi
                    while hi>=0 and nums[hi]==nums[p]: hi-=1
                elif nums[lo]+nums[hi]>-1*nums[i]:
                    p=hi
                    while hi>=0 and nums[hi]==nums[p]: hi-=1
                elif nums[lo]+nums[hi]<-1*nums[i]:
                    p=lo
                    while lo<l and nums[lo]==nums[p]: lo+=1
            p=i
            while i<l and nums[p]==nums[i]: i+=1
        return ans
                    
                    
        